docs/references/architecture/adr-113-mempool-lanes.md

```diff
@@ -294,254 +296,285 @@ TODO
+sends an `Info` request to fetch the app information, and then replays any stored block needed to
+sync CometBFT with the app. The lane information is needed regardless of whether state sync is
+enabled, so one option is to query the app information outside of the `Handshaker`.
+
+In this proposed approach, updating the lane definitions will require a single governance proposal
+for updating the software. CometBFT will not need to deal with dynamic lane changes: it will just
+need to set up the lanes when starting up (whether afresh or in recovery mode).
+
+Different nodes also need to agree on the lanes they use. When a node connects to a peer, they both
+perform a handshake to agree on some basic information (see `DefaultNodeInfo`), including the
+version of the application they are executing. Since the application includes the lane definitions,
+both nodes will also agree on the lanes they implement. Therefore we don't need to modify the P2P
+handshake.
+
+### Internal data structures
+
+In the mempool, a lane is defined by its priority:
+```golang
+type Lane uint32
+```
+
+Currently, the `CListMempool` data structure has two fields to store and access transactions:
+```golang
+txs    *clist.CList // Concurrent list of mempool entries.
+txsMap sync.Map     // Map of type TxKey -> *clist.CElement, for quick access to elements in txs.
+```
+
+With the introduction of lanes, the main change will be to divide the `CList` data structure into
+$N$ `Clist`s, one per lane. `CListMempool` will have the following fields:
+```golang
+lanes   map[Lane]*clist.CList
+txsMap  sync.Map // Map of type TxKey -> *clist.CElement, for quick access to elements in lanes.
+txLanes sync.Map // Map of type TxKey -> Lane, for quick access to the lane corresponding to a tx.
+
+// Fixed variables set during initialization.
+defaultLane Lane
+sortedLanes []Lane // Lanes sorted by priority
+```
+The auxiliary fields `txsMap` and `txLanes` are, respectively, for direct access to the mempool
+entries, and for direct access to the lane of a given transaction.
+
+If the application does not implement lanes (that is, it responds with empty values in
+`InfoResponse`), then `defaultLane` will be set to 1, and `lanes` will have only one entry for the
+default lane.
+
+`CListMempool` also contains the cache, which is only needed before transactions have a lane
+assigned. Since the cache is independent of the lanes, we do not need to modify it.
+
+### Configuration
+
+For an MVP, all lanes can share the same mempool configuration. In this scenario, all lanes will be
+capped by the total mempool capacities as currently defined in the configuration. Namely, these are:
+- `Size`, the total number of transactions allowed in the mempool, 
+- `MaxTxsBytes`, the maximum total number of bytes of the mempool, and
+- `MaxTxBytes`, the maximum size in bytes of a single transaction accepted into the mempool.
+
+With lanes, the total size of the mempool will be the sum of the sizes of all lanes.
```

Comment:
Do we want to divide the total size evenly across lanes? (or are we gonna expunge a TX from a lower lane when a full mempool receives a TX?)

Reply:
I'd say it's not needed for the MVP but it could be an interesting improvement for the future. Equal sizes for all lanes may not make sense when lanes are defined by transaction size. For example, a lane for only very big txs. Probably we won't see this issue when we add custom configurations per lane.